### ISSUE TITLE
configure firewall inside container

### ISSUE THREAD

**Issue by M1cha** [containers/netavark#528]
I don't know if docker or CNI offer a solution to this but this is my usecase:
- I'm running a small homeserver with tons of containers. Each accessible container has it's own `.home.arpa` domain statically assigned by the router.
- I don't want these containers to be able to access or even see anything in my home network. Most containers are not even allowed to use the internet.
- I still want to access them as efficiently(in terms of throughput) as possible.
- The currently solution is to use the default-bridge, `--publish` certain ports, and install nftables rules on the host that block all outgoing traffic for the containers mac(except for conntracking). The downside is that I have to set all `.home.arpa` domains to the same IP(the one of the host) and I have to use different ports to prevent conflicts. That's not as usable as using port 443 for everything.
- big production servers simply have an ingress (e.g. nginx) server which routes traffic based on the domain.
  - On low-power ARM devices that drastically reduces throughput
  - works with HTTP only
- you could use l3s ipvlan but that reduces performance as well

AFAICT the most efficient solution would be to give every container a macvlan interface. Unfortunately  you can't filter macvlan traffic from the host. You obviously don't want to give containers NET_ADMIN since that would allow them to bypass all solutions mentioned previously.

TL;DR:
**So what's needed is a way to make podman/netavark/crun set up arbitrary user-provided nftables before dropping privileges.**

I am willing to implement this myself but we need to come up with a good way to do this.

**Comment by M1cha** [comment]
1) yes, I think this is out of scope for netvark
2) I solved this in a much simpler way that doesn't require the oci-hook to know the IP config: by defining an IP range in the same network that the routers DHCP doesn't allocate and writing an [arp-reply daemon](https://github.com/M1cha/arp-reply) so I receive traffic for those IPs without having to assign them all to the podman-hosts interface.
3) In case you wanna go with a hook anyway, here's my previous code:

firewall.json:
```yaml
{
  "version": "1.0.0",
  "hook": {
    "path": "/opt/container-files/hooks/scripts/firewall.sh"
  },
  "when": {
    "annotations": {
      "^homeserver\\.firewall\\..*": ".*"
    }
  },
  "stages": ["createRuntime"]
}
```

firewall.sh:
```bash
#!/bin/bash

set -euo pipefail

state=$(</dev/stdin)
pid=$(jq -re .pid <<< "$state")

get_config() {
    set +e
    output=$(jq -re ".annotations[\"homeserver.firewall.$1\"]" <<< "$state")
    code="$?"
    set -e

    if [ "$code" == "0" ]; then
        echo -n "$output"
    else
        # default value
        echo -n "${2:-}"
    fi
}

enabled=$(get_config enabled)
if [ "$enabled" != "1" ]; then
    # why where we called then?
    exit 1
fi

nsenter -t "$pid" -n nft -f /dev/stdin <<EOF
# your rules
EOF
```
